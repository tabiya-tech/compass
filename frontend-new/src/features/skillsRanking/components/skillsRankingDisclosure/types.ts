export const jobSeekerComparisonLabels = ["LOWEST", "SECOND LOWEST", "MIDDLE", "SECOND HIGHEST", "HIGHEST"]