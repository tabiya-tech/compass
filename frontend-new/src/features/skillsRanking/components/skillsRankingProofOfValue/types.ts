export enum EffortType {
  TIME_BASED = "time_based",
  WORK_BASED = "work_based",
}