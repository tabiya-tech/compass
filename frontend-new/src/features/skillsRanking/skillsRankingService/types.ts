export type SkillsRankingMetrics = {
  cancelled_after?: string,
  succeeded_after?:string,
  puzzles_solved?:number,
  correct_rotations?:number,
  clicks_count?:number,
}