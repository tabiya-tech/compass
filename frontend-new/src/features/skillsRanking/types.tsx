export enum CompareAgainstGroup {
  AGAINST_OTHER_JOB_SEEKERS = "against_other_job_seekers",
  AGAINST_JOB_MARKET = "against_job_market"
}

export enum ButtonOrderGroup {
  SKIP_BUTTON_FIRST = "skip_button_first",
  VIEW_BUTTON_FIRST = "view_button_first"
}

export interface SkillRankingExperimentGroups {
  compare_against: CompareAgainstGroup;
  button_order: ButtonOrderGroup;
  delayed_results: boolean;
}

export enum SkillsRankingPhase {
  INITIAL = "INITIAL",
  SKIPPED = "SKIPPED",
  CANCELLED = "CANCELLED",
  SELF_EVALUATING = "SELF_EVALUATING",
  EVALUATED = "EVALUATED"
}

export interface SkillsRankingState {
  session_id: number;
  experiment_groups: SkillRankingExperimentGroups;
  phase: SkillsRankingPhase;
  ranking: string | null;
  self_ranking: string | null;
}

export const skillsRankingStateDefault: SkillsRankingState = {
  session_id: -1,
  experiment_groups: {
    compare_against: CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS,
    button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
    delayed_results: false,
  },
  phase: SkillsRankingPhase.INITIAL,
  ranking: null,
  self_ranking: null,
}

export interface SkillsRankingResult {
  ranking: string;
}