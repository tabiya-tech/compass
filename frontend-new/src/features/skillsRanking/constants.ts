import { SkillsRankingService } from "./skillsRankingService/skillsRankingService";

export const SKILLS_RANKING_FEATURE_ID = "4b0c7428-9c01-4688-81fd-d3ef159bce79";

// Typing durations for different components
export const TYPING_DURATION_MS = 5000;
export const MESSAGE_DURATION_MS = 5000;
export const CALCULATION_DELAY = 60000;
export const EFFORT_METRICS_UPDATE_INTERVAL = 2000;

// Puzzle constants
export const DEFAULT_STRINGS = ["GJRLK", "FQZNC", "EKJGR", "CJFLQ", "GRKLE"];

// Config constants - these values are unlikely to change during runtime
export const getJobPlatformUrl = () => SkillsRankingService.getInstance().getConfig().config.jobPlatformUrl;
export const getAirtimeBudget = () => SkillsRankingService.getInstance().getConfig().config.airtimeBudget;
