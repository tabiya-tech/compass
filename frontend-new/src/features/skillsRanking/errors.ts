
export class SkillsRankingError extends Error {
    constructor(message: string, cause?: unknown) {
      super(message);
      this.name = "SkillsRankingError";
      this.cause = cause;
    }
  }