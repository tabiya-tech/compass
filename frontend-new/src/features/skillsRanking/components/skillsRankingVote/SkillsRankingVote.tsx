import React, { useContext, useEffect, forwardRef } from "react";
import { Box, Typography, useTheme } from "@mui/material";
import { SkillsRankingPhase, SkillsRankingState, CompareAgainstGroup } from "src/features/skillsRanking/types";
import { MessageContainer } from "src/chat/chatMessage/compassChatMessage/CompassChatMessage";
import ChatBubble from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";
import SkillsRankingVoteItem from "./components/SkillsRankingVoteItem";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

const uniqueId = "cb5e9e43-c288-4597-b6f4-a268623c9e61";

export const SKILLS_RANKING_VOTE_MESSAGE_TYPE = `skills-ranking-vote-message-${uniqueId}`;

export const DATA_TEST_ID = {
  SKILLS_RANKING_VOTE_CONTAINER: `skills-ranking-vote-container-${uniqueId}`,
  SKILLS_RANKING_VOTE_TEXT: `skills-ranking-vote-text-${uniqueId}`,
  SKILLS_RANKING_VOTE_ICON: `skills-ranking-vote-icon-${uniqueId}`,
  SKILLS_RANKING_VOTE_ERROR: `skills-ranking-vote-error-${uniqueId}`,
};

export interface SkillsRankingVoteProps {
  message: string;
  onRankSelect: (rank: string) => Promise<void>;
  disabled?: boolean;
  skillsRankingState: SkillsRankingState;
}

export const OPTIONS = [
  { value: "0%", percent: 0 },
  { value: "10%", percent: 10 },
  { value: "20%", percent: 20 },
  { value: "30%", percent: 30 },
  { value: "40%", percent: 40 },
  { value: "50%", percent: 50 },
  { value: "60%", percent: 60 },
  { value: "70%", percent: 70 },
  { value: "80%", percent: 80 },
  { value: "90%", percent: 90 },
  { value: "100%", percent: 100 },
];

export const QUESTION_TEXTS = {
  [CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS]:
    "Before I tell you, how do your current skillset ranks compared to other job seekers?",
  [CompareAgainstGroup.AGAINST_JOB_MARKET]:
    "Before I tell you, how do your current skillset ranks compared to skills demanded in available jobs?",
};

const SkillsRankingVote = forwardRef<HTMLDivElement, SkillsRankingVoteProps>(
  ({ skillsRankingState, message, onRankSelect, disabled = false }, ref) => {
    const theme = useTheme();
    const isOnline = useContext(IsOnlineContext);
    const [questionText, setQuestionText] = React.useState<string>("");
    const [selectedIndex, setSelectedIndex] = React.useState<number | null>(null);
    const [isRankSelected, setIsRankSelected] = React.useState<boolean>(false);
    const [hoveredIndex, setHoveredIndex] = React.useState<number | null>(null);

    useEffect(() => {
      setQuestionText(QUESTION_TEXTS[skillsRankingState.experiment_groups.compare_against]);
      if (skillsRankingState.phase === SkillsRankingPhase.SELF_EVALUATING) {
        setSelectedIndex(null);
        setIsRankSelected(false);
      } else {
        setIsRankSelected(true);
        if (skillsRankingState.self_ranking !== null) {
          setSelectedIndex(
            OPTIONS.findIndex((option) => option.percent === parseInt(skillsRankingState.self_ranking as string))
          );
        } else {
          setSelectedIndex(null);
        }
      }
    }, [
      skillsRankingState.phase,
      skillsRankingState.experiment_groups.compare_against,
      skillsRankingState.self_ranking,
    ]);

    // Disable buttons if we're not in the self-evaluating state
    const isDisabled = disabled || isRankSelected;

    const handleRankSelect = (rank: string) => {
      setSelectedIndex(OPTIONS.findIndex((option) => option.value === rank));
      setIsRankSelected(true);
      onRankSelect(rank);
    };

    return (
      <Box ref={ref}>
        <MessageContainer
          origin={ConversationMessageSender.COMPASS}
          data-testid={DATA_TEST_ID.SKILLS_RANKING_VOTE_CONTAINER}
        >
          <ChatBubble message={message} sender={ConversationMessageSender.COMPASS}>
            <Box display="flex" flexDirection="column" gap={theme.fixedSpacing(theme.tabiyaSpacing.sm)}>
              <Typography variant="body1" data-testid={DATA_TEST_ID.SKILLS_RANKING_VOTE_TEXT}>
                {questionText}
              </Typography>
              {/* Options */}
              <Box
                display="flex"
                alignItems="center"
                width="100%"
                sx={{ marginY: theme.fixedSpacing(theme.tabiyaSpacing.md) }}
              >
                {OPTIONS.map((option, index) => (
                  <SkillsRankingVoteItem
                    key={option.value}
                    option={option}
                    index={index}
                    selectedIndex={selectedIndex}
                    hoveredIndex={hoveredIndex}
                    disabled={!isOnline || isDisabled}
                    onSelect={(index: number) => {
                      setSelectedIndex(index);
                      handleRankSelect(option.value);
                    }}
                    onHover={setHoveredIndex}
                    isLast={index === OPTIONS.length - 1}
                    showTopLabel={index % 2 === 0}
                    showBottomLabel={index % 2 !== 0}
                  />
                ))}
              </Box>
            </Box>
          </ChatBubble>
        </MessageContainer>
      </Box>
    );
  }
);

export default SkillsRankingVote;
