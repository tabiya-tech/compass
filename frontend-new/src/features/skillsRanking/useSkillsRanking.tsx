import { useCallback, useEffect, useRef, useState } from "react";
import { SkillsRankingPhase, SkillsRankingState, skillsRankingStateDefault } from "src/features/skillsRanking/types";
import { SkillsRankingService } from "src/features/skillsRanking/skillsRankingService/skillsRankingService";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { IChatMessage } from "src/chat/Chat.types";
import SkillsRankingPrompt, {
  SKILLS_RANKING_PROMPT_MESSAGE_TYPE,
  SkillsRankingPromptProps,
} from "src/features/skillsRanking/components/skillsRankingPrompt/SkillsRankingPrompt";
import SkillsRankingVote, {
  SKILLS_RANKING_VOTE_MESSAGE_TYPE,
  SkillsRankingVoteProps,
} from "src/features/skillsRanking/components/skillsRankingVote/SkillsRankingVote";
import SkillsRankingResult, {
  SKILLS_RANKING_RESULT_MESSAGE_TYPE,
  SkillsRankingResultProps,
} from "src/features/skillsRanking/components/skillsRankingResult/SkillsRankingResult";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import { SkillsRankingError } from "./errors";
import CancellableTypingChatMessage, {
  CANCELABLE_TYPING_CHAT_MESSAGE_TYPE,
  CancellableTypingChatMessageProps,
} from "./components/cancellableTypingChatMessage/CancellableTypingChatMessage";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { generateTypingMessage } from "src/chat/util";
import { TypingChatMessageProps } from "src/chat/chatMessage/typingChatMessage/TypingChatMessage";

type SkillsRankingMessageProps = SkillsRankingPromptProps | SkillsRankingVoteProps | SkillsRankingResultProps | CancellableTypingChatMessageProps;

const TYPING_TIMEOUT = 5000; // 5 seconds
const CANCELABLE_TYPING_TIMEOUT = 10000; // 10 seconds
const DELAY_AFTER_RESULTS = 3000; // 3 seconds

const SkillsRankingMessageIds = {
  [SKILLS_RANKING_PROMPT_MESSAGE_TYPE]: "skills-ranking-prompt-message",
  [SKILLS_RANKING_VOTE_MESSAGE_TYPE]: "skills-ranking-vote-message",
  [SKILLS_RANKING_RESULT_MESSAGE_TYPE]: "skills-ranking-result-message",
  [CANCELABLE_TYPING_CHAT_MESSAGE_TYPE]: "skills-ranking-cancellable-typing-message",
} as const;

export const useSkillsRanking = (addMessage: (message: IChatMessage<any>) => void, removeMessage: (messageId: string) => void) => {
  const pendingTimeouts = useRef<{ [key: string]: NodeJS.Timeout }>({});
  const [currentSkillsRankingState, setCurrentSkillsRankingState] = useState<SkillsRankingState>(skillsRankingStateDefault);
  const [ranking, setRanking] = useState<string>("");
  const [isLoading, setIsLoading] = useState(false);
  const abortControllerRef = useRef<AbortController | null>(null);
  const onFinishRef = useRef<(() => void) | null>(null);
  const lastHandledState = useRef<SkillsRankingPhase | null>(null);
  const shownMessages = useRef<Set<string>>(new Set());
  const { enqueueSnackbar } = useSnackbar();

  const handleError = useCallback((error: Error) => {
    console.error(new SkillsRankingError("Error in skills ranking flow:", error));
    enqueueSnackbar("An error occurred while calculating your skills ranking. Please try again later.", {
      variant: "error",
    });
  }, [enqueueSnackbar]);

  const clearAllPendingTimeouts = useCallback(() => {
    Object.values(pendingTimeouts.current).forEach(timeout => clearTimeout(timeout));
    pendingTimeouts.current = {};
  }, []);

  const createMessage = useCallback(<T extends SkillsRankingMessageProps>(type: string, props: T, Component: React.ComponentType<T>): IChatMessage<T> => ({
    message_id: SkillsRankingMessageIds[type as keyof typeof SkillsRankingMessageIds],
    type,
    payload: props,
    sender: ConversationMessageSender.COMPASS,
    component: (props: T) => <Component {...props} />,
  }), []);

  const fetchRanking = useCallback(async (signal: AbortSignal) => {
    setIsLoading(true);
    try {
      const sessionId = UserPreferencesStateService.getInstance().getActiveSessionId();
      if (sessionId === null) {
        throw new Error("No active session found");
      }
      const skillsRankingService = SkillsRankingService.getInstance();
      const result = await skillsRankingService.getRanking(sessionId, signal);
      setRanking(result.ranking);
      return result.ranking;
    } catch (error) {
      if (error instanceof Error && error.name === 'AbortError') {
        console.log('Ranking fetch was aborted');
        return null;
      }
      const errorMessage = error instanceof Error ? error.message : "Failed to fetch ranking";
      handleError(error instanceof Error ? error : new Error(errorMessage));
      return null;
    } finally {
      setIsLoading(false);
    }
  }, [handleError]);

  const handleStateTransition = useCallback(async (newState: SkillsRankingPhase, rank?: string) => {
    const sessionId = UserPreferencesStateService.getInstance().getActiveSessionId();
    if (!sessionId) return;
    
    try {
      const state = await SkillsRankingService.getInstance().updateSkillsRankingState(
        sessionId,
        newState,
        rank
      );
      setCurrentSkillsRankingState(state);
      return state;
    } catch (error) {
      handleError(error as Error);
      return null;
    }
  }, [handleError]);

  const handleStateMessages = useCallback((state: SkillsRankingState, onFinish: () => void) => {
    if (lastHandledState.current === state.phase) return;
    lastHandledState.current = state.phase;

    const showTypingAndThen = (callback: () => void) => {
      let typingMsg: IChatMessage<TypingChatMessageProps> | null = null;

      if (state.phase !== SkillsRankingPhase.SKIPPED && state.phase !== SkillsRankingPhase.CANCELLED && state.phase !== SkillsRankingPhase.EVALUATED) {
        typingMsg = generateTypingMessage();
        addMessage(typingMsg);
      }

      pendingTimeouts.current['showMessage'] = setTimeout(() => {
        if (typingMsg) {
          removeMessage(typingMsg.message_id);
        }
        callback();
      }, typingMsg ? TYPING_TIMEOUT : 0);
    };

    const handleState = () => {
      switch (state.phase) {
        case SkillsRankingPhase.INITIAL:
          if (!shownMessages.current.has(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE])) {
            addMessage(createMessage(SKILLS_RANKING_PROMPT_MESSAGE_TYPE, {
              message: "Please rate your skills",
              onView: async () => {
                const newState = await handleStateTransition(SkillsRankingPhase.SELF_EVALUATING);
                if (newState) handleStateMessages(newState, onFinish);
              },
              onSkip: async () => {
                removeMessage(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE]);
                const newState = await handleStateTransition(SkillsRankingPhase.SKIPPED);
                if (newState) handleStateMessages(newState, onFinish);
              },
              disabled: false,
              skillsRankingState: state,
            }, SkillsRankingPrompt));
            shownMessages.current.add(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE]);
          }
          break;

        case SkillsRankingPhase.SELF_EVALUATING:
          if (!shownMessages.current.has(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE])) {
            addMessage(createMessage(SKILLS_RANKING_PROMPT_MESSAGE_TYPE, {
              message: "Please rate your skills",
              onView: async () => {
                const newState = await handleStateTransition(SkillsRankingPhase.SELF_EVALUATING);
                if (newState) handleStateMessages(newState, onFinish);
              },
              onSkip: async () => {
                removeMessage(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE]);
                const newState = await handleStateTransition(SkillsRankingPhase.SKIPPED);
                if (newState) handleStateMessages(newState, onFinish);
              },
              disabled: true,
              skillsRankingState: state,
            }, SkillsRankingPrompt));
            shownMessages.current.add(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE]);
          }
          
          if (!shownMessages.current.has(SkillsRankingMessageIds[SKILLS_RANKING_VOTE_MESSAGE_TYPE])) {
            addMessage(createMessage(SKILLS_RANKING_VOTE_MESSAGE_TYPE, {
              message: "Please rate your skills",
              onRankSelect: async (rank: string) => {
                // Show typing message first
                const cancelableTypingMsg = createMessage(CANCELABLE_TYPING_CHAT_MESSAGE_TYPE, {
                  message: "Calculating your skills ranking",
                  onCancel: async () => {
                    clearAllPendingTimeouts();
                    if (abortControllerRef.current) {
                      abortControllerRef.current.abort();
                      abortControllerRef.current = null;
                    }
                    removeMessage(SkillsRankingMessageIds[CANCELABLE_TYPING_CHAT_MESSAGE_TYPE]);
                    const newState = await handleStateTransition(SkillsRankingPhase.CANCELLED);
                    if (newState) handleStateMessages(newState, onFinish);
                  },
                }, CancellableTypingChatMessage);
                addMessage(cancelableTypingMsg);

                // Start fetching ranking
                abortControllerRef.current = new AbortController();
                await fetchRanking(abortControllerRef.current.signal);

                // Wait for typing timeout before transitioning state
                pendingTimeouts.current['rankSelect'] = setTimeout(async () => {
                  removeMessage(cancelableTypingMsg.message_id);
                  const newState = await handleStateTransition(SkillsRankingPhase.EVALUATED, rank);
                  if (newState) handleStateMessages(newState, onFinish);
                }, CANCELABLE_TYPING_TIMEOUT);
              },
              skillsRankingState: state,
            }, SkillsRankingVote));
            shownMessages.current.add(SkillsRankingMessageIds[SKILLS_RANKING_VOTE_MESSAGE_TYPE]);
          }
          break;

        case SkillsRankingPhase.EVALUATED:
          if (!shownMessages.current.has(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE])) {
            addMessage(createMessage(SKILLS_RANKING_PROMPT_MESSAGE_TYPE, {
              message: "Please rate your skills",
              onView: async () => {
                const newState = await handleStateTransition(SkillsRankingPhase.SELF_EVALUATING);
                if (newState) handleStateMessages(newState, onFinish);
              },
              onSkip: async () => {
                removeMessage(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE]);
                const newState = await handleStateTransition(SkillsRankingPhase.SKIPPED);
                if (newState) handleStateMessages(newState, onFinish);
              },
              disabled: true,
              skillsRankingState: state,
            }, SkillsRankingPrompt));
            shownMessages.current.add(SkillsRankingMessageIds[SKILLS_RANKING_PROMPT_MESSAGE_TYPE]);
          }
          
          if (!shownMessages.current.has(SkillsRankingMessageIds[SKILLS_RANKING_VOTE_MESSAGE_TYPE])) {
            addMessage(createMessage(SKILLS_RANKING_VOTE_MESSAGE_TYPE, {
              message: "Please rate your skills",
              onRankSelect: async (rank: string) => {
                const newState = await handleStateTransition(SkillsRankingPhase.EVALUATED, rank);
                if (newState) handleStateMessages(newState, onFinish);
              },
              skillsRankingState: state,
            }, SkillsRankingVote));
            shownMessages.current.add(SkillsRankingMessageIds[SKILLS_RANKING_VOTE_MESSAGE_TYPE]);
          }
          console.log('state:', state);
          
          if (!shownMessages.current.has(SkillsRankingMessageIds[SKILLS_RANKING_RESULT_MESSAGE_TYPE])) {
            addMessage(createMessage(SKILLS_RANKING_RESULT_MESSAGE_TYPE, {
              message: "Here's what we found",
              skillsRankingState: state,
              isLoading,
            }, SkillsRankingResult));
            shownMessages.current.add(SkillsRankingMessageIds[SKILLS_RANKING_RESULT_MESSAGE_TYPE]);
          }
          
          const finishTimeoutId = setTimeout(() => {
            onFinish();
          }, DELAY_AFTER_RESULTS);
          
          pendingTimeouts.current['finish'] = finishTimeoutId;
          break;

        case SkillsRankingPhase.SKIPPED:
        case SkillsRankingPhase.CANCELLED:
          console.debug("SKIPPED or CANCELLED, calling onFinish");
          onFinish();
          break;
      }
    };

    showTypingAndThen(handleState);
  }, [addMessage, removeMessage, createMessage, handleStateTransition, clearAllPendingTimeouts, fetchRanking, isLoading]);

  const showSkillsRanking = useCallback(async (onFinish: () => void) => {
    clearAllPendingTimeouts();
    shownMessages.current.clear(); // Reset shown messages when starting a new flow
    try {
      const sessionId = UserPreferencesStateService.getInstance().getActiveSessionId();
      if (!sessionId) throw new Error("No active session found");
      
      let state = await SkillsRankingService.getInstance().getSkillsRankingState(sessionId);
      // If no state exists, create one with INITIAL phase
      state ??= await SkillsRankingService.getInstance().updateSkillsRankingState(
        sessionId,
        SkillsRankingPhase.INITIAL,
      );
      setCurrentSkillsRankingState(state);
      onFinishRef.current = onFinish;
      handleStateMessages(state, onFinish);
    } catch (error) {
      console.error("Error in skills ranking flow:", error);
      onFinish();
    }
  }, [clearAllPendingTimeouts, handleStateMessages]);

  useEffect(() => {
    if (onFinishRef.current) {
      handleStateMessages(currentSkillsRankingState, onFinishRef.current);
      onFinishRef.current = null;
    }
  }, [ranking, isLoading, currentSkillsRankingState.phase, handleStateMessages, currentSkillsRankingState]);

  return { showSkillsRanking };
}; 