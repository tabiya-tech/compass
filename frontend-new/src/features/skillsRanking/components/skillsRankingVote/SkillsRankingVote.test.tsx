// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import SkillsRankingVote, {
  DATA_TEST_ID,
  QUESTION_TEXTS,
  OPTIONS,
} from "src/features/skillsRanking/components/skillsRankingVote/SkillsRankingVote";
import { CompareAgainstGroup, ButtonOrderGroup } from "src/features/skillsRanking/types";
import { SkillsRankingPhase, SkillsRankingState } from "src/features/skillsRanking/types";
import { DATA_TEST_ID as CHAT_BUBBLE_DATA_TEST_ID } from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import userEvent from "@testing-library/user-event";

// mock SkillsRankingVoteItem component
jest.mock("src/features/skillsRanking/components/skillsRankingVote/components/SkillsRankingVoteItem", () => ({
  __esModule: true,
  default: (props: any) => (
    <div data-testid="skills-ranking-vote-item">
      <div data-testid="skills-ranking-vote-item-radio" onClick={() => props.onSelect?.(props.index)} />
      {!props.isLast && <div data-testid="skills-ranking-vote-item-divider" />}
    </div>
  ),
}));

describe("SkillsRankingVote tests", () => {
  describe("render tests", () => {
    test("should render component successfully", () => {
      // GIVEN a skills ranking state
      const givenSkillsRankingState: SkillsRankingState = {
        phase: SkillsRankingPhase.SELF_EVALUATING,
        experiment_groups: {
          compare_against: CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS,
          delayed_results: false,
          button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
        },
        session_id: 1,
        ranking: "",
        self_ranking: null,
      };

      // WHEN the component is rendered
      render(
        <SkillsRankingVote
          message="Please rate your skills"
          onRankSelect={jest.fn()}
          skillsRankingState={givenSkillsRankingState}
        />
      );

      // THEN expect the container to be in the document
      const container = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_VOTE_CONTAINER);
      expect(container).toBeInTheDocument();
      // AND expect the message bubble to be visible
      expect(screen.getByTestId(CHAT_BUBBLE_DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER)).toBeInTheDocument();
      // AND expect the text to be in the document
      expect(screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_VOTE_TEXT)).toBeInTheDocument();
      // AND the SkillsRankingVoteItem to be visible
      const voteItems = screen.getAllByTestId("skills-ranking-vote-item");
      expect(voteItems).toHaveLength(OPTIONS.length);
      // AND to match the snapshot
      expect(container).toMatchSnapshot();
      // AND no errors or warnings to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
    });

    test.each([
      [CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS, QUESTION_TEXTS[CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS]],
      [CompareAgainstGroup.AGAINST_JOB_MARKET, QUESTION_TEXTS[CompareAgainstGroup.AGAINST_JOB_MARKET]],
    ])("should render correct question text for %s", (experimentGroup, expectedText) => {
      // GIVEN a skills ranking state
      const givenSkillsRankingState: SkillsRankingState = {
        phase: SkillsRankingPhase.SELF_EVALUATING,
        experiment_groups: {
          compare_against: experimentGroup,
          button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
          delayed_results: false,
        },
        session_id: 1,
        ranking: "",
        self_ranking: null,
      };

      // WHEN the component is rendered
      render(
        <SkillsRankingVote
          message="Please rate your skills"
          onRankSelect={jest.fn()}
          skillsRankingState={givenSkillsRankingState}
        />
      );

      // THEN expect the correct question text to be in the document
      const questionText = screen.getByTestId(DATA_TEST_ID.SKILLS_RANKING_VOTE_TEXT);
      expect(questionText).toBeInTheDocument();
      expect(questionText).toHaveTextContent(expectedText);
      // AND no errors or warnings to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
    });
  });

  describe("action tests", () => {
    test("should call onRankSelect when a rank icon is clicked", async () => {
      // GIVEN a skills ranking state
      const givenSkillsRankingState: SkillsRankingState = {
        phase: SkillsRankingPhase.SELF_EVALUATING,
        experiment_groups: {
          compare_against: CompareAgainstGroup.AGAINST_OTHER_JOB_SEEKERS,
          button_order: ButtonOrderGroup.SKIP_BUTTON_FIRST,
          delayed_results: false,
        },
        session_id: 1,
        ranking: "",
        self_ranking: null,
      };
      // AND onRankSelect mock function
      const onRankSelect = jest.fn();

      // WHEN the component is rendered
      render(
        <SkillsRankingVote
          message="Please rate your skills"
          onRankSelect={onRankSelect}
          skillsRankingState={givenSkillsRankingState}
        />
      );

      // AND the first rank icon is clicked
      const rankIcon = screen.getAllByTestId("skills-ranking-vote-item-radio")[0];
      await userEvent.click(rankIcon);

      // THEN expect the onRankSelect function to have been called with the correct argument
      expect(onRankSelect).toHaveBeenCalledWith(OPTIONS[0].value);
      // AND no errors or warnings to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
    });

    test.todo("should be disabled when not in self-evaluating state");
  });
});
