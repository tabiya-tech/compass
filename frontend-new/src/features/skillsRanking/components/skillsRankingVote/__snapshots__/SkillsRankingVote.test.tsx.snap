// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillsRankingVote tests render tests should render component successfully 1`] = `
<div
  class="MuiBox-root css-1c8zvho"
  data-testid="skills-ranking-vote-container-cb5e9e43-c288-4597-b6f4-a268623c9e61"
  origin="COMPASS"
>
  <div
    class="MuiBox-root css-u9ttk8"
    data-testid="chat-message-bubble-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
    origin="COMPASS"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-13zm2uq-MuiTypography-root"
      data-testid="chat-message-bubble-message-text-6e685eeb-2b54-432a-8b66-8a81633b3981"
    >
      Please rate your skills
    </p>
    <div
      class="MuiBox-root css-0"
      data-testid="chat-message-bubble-message-footer-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
    >
      <div
        class="MuiBox-root css-1821gv5"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1czz4sp-MuiTypography-root"
          data-testid="skills-ranking-vote-text-cb5e9e43-c288-4597-b6f4-a268623c9e61"
        >
          Before I tell you, how do your current skillset ranks compared to other job seekers?
        </p>
        <div
          class="MuiBox-root css-1a23d07"
        >
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
            <div
              data-testid="skills-ranking-vote-item-divider"
            />
          </div>
          <div
            data-testid="skills-ranking-vote-item"
          >
            <div
              data-testid="skills-ranking-vote-item-radio"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
